### DISCUSSION TITLE
Continuous Deployment Events

### DISCUSSION THREAD

**Discussion by sbtaylor15** [cdfoundation/sig-events#28] [General]
What is an environment?  
- Prod has 100 clusters.  Is an individual cluster an environment or is the cluster endpoint in an environment?

Should services and environments have starting and completed events?

Do deployments events have sub-events, ie for deployments to each cluster would be a sub-event of the higher level deployment event?

**Comment by erkist** [comment]
> What is an environment?

Without going into the other parts part of the questions, which is outside of my domain, we have one-or-two definitions of Environment floating around, I believe.

In the Continuous Deployment part of the spec, an environment is something that is (pretty frequently, I assume) created/removed and that can run services. I take this to mean e.g. a Kubernetes cluster or a Docker host or similar.

Eiffel has a more generic concept to cover both deployment and test environments: [it may be a virtual image, a complete mechatronic system of millions of independent parts, or anything in between](https://github.com/eiffel-community/eiffel/blob/master/eiffel-vocabulary/EiffelEnvironmentDefinedEvent.md).

ESCI(D) is mostly concerned with environments on which activities are run (mainly test activities), and refers to to said environments as [resource sets](https://github.com/erkist/sig-interoperability/blob/erkist-esci/workstreams/events_in_cicd/input/enterprise-scale-ci/README.md#resource-set) (in ESCI, the term "environment" is used to differentiate between e.g. "production" and "staging", not between individual resource sets inside said environment, e.g. "prod-1" and "prod-2").

**Reply by sbtaylor15** [reply]
I view an Environment as a way to group resources together for a specific purpose, Development, QA Testing, or Production.  You can have as many Environments as needed.  Also, they can be transient, for example QA-7 may exist for specific feature testing and then be destroyed when testing is complete.

**Reply by m-linner-ericsson** [reply]
Have we stated why you want to send an environment event? Is it:
* To be able to link to, e.g. "tests were run on env x", "artifact was deployed on env y"
* To notify others that an environment is ready to use?
* Just for keeping record e.g. visualization and metrics
